### ISSUE TITLE
Scheduler: use `Pending` state in in-tree plugins

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#123227]
/sig scheduling
/kind feature

`Pending` status is newly introduced for a efficient requeueing, see the comment to understand this:
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/interface.go#L114-L130

~~For now, only DRA uses this status~~, but we can utilize this for other in-tree plugins as well. (EDIT: they used, but no longer now after they moved to the structured parameters)
For example, volume binding probably can return `Pending` when Pod's PVC is not yet created. There should be more usecase everywhere.

**Comment by kerthcet** [comment]
There's another problem we have here is Pending will run the preemption, however, it's unresolvable with preemption. We may need status like `PendingAndUnresolvable` to skip the preemption.

/assign
Let me take a look.

**Comment by sanposhiho** [comment]
Thanks,

>  it's unresolvable with preemption sometimes

Or, maybe always? If all the usecases of Pending don't need a preempttion seemingly, we can just change Pending not to trigger preemption.

**Comment by YamasouA** [comment]
@kerthcet
Hi!
Do you have any progress on it? If not, Can I take over this issue?

**Comment by kerthcet** [comment]
Please go ahead.
/unassign

**Comment by YamasouA** [comment]
@kerthcet 
Thankyou!!

**Comment by sanposhiho** [comment]
@YamasouA Any update on this?

**Comment by saza-ku** [comment]
> For now, only DRA uses this status,

I can't find DRA's codes that uses `Pending`. I guess that was added in #119517 but removed in #128003.

https://github.com/kubernetes/kubernetes/pull/128003/files#diff-cf6f598b41628749568815f4aec92078e3a1e4ce07020913f080e7181aa11163L1572-L1586

**Comment by sanposhiho** [comment]
Yes, they already removed the use at the transition to the structured parameters.

**Comment by saza-ku** [comment]
I see.

I read #119517 to understand when to return `Pending`.

> SuccessButReject is used when the scheduler succeeds everything in the scheduling cycle but needs to reject Pod once, for example, in order to wait for the external component to do something.

The volume binding plugin triggers volume provisioning and wait for it in PreBind phase when a pvc is not provisioned yet. We could replace it with just triggering volume provisioning and returning Pending.

https://github.com/kubernetes/kubernetes/blob/1c32094c03f721cd3dce49c9743f1e7f3e564b13/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L516

https://github.com/kubernetes/kubernetes/blob/1c32094c03f721cd3dce49c9743f1e7f3e564b13/pkg/scheduler/framework/plugins/volumebinding/binder.go#L466

@YamasouA May I work on it if you have no progress?

**Comment by sanposhiho** [comment]
> We could replace it with just triggering volume provisioning and returning Pending.

How does it benefit the performance? PreBind is running within goroutine and hence the waiting of the provision doesn't block other scheduling cycles currently.

**Comment by saza-ku** [comment]
> PreBind is running within goroutine and hence the waiting of the provision doesn't block other scheduling cycles currently.

Right. It doesn't affect the performance.

**Comment by sanposhiho** [comment]
..then do we have any benefit converting it to use `Pending` status?

**Comment by saza-ku** [comment]
We　don't, so we may leave it.

**Comment by YamasouA** [comment]
@saza-ku 

> May I work on it if you have no progress?

Yes. Please🙇‍♂️

**Comment by saza-ku** [comment]
Thanks!

/assign

**Comment by saza-ku** [comment]
I'll first work on cases where a Pod's PVCs are not found in plugins related to volumes as mentioned in the issue description.

**Comment by sanposhiho** [comment]
> For example, volume binding probably can return Pending when Pod's PVC is not yet created.

Sorry, wait a sec about this. The previous silly me said this though, the current smarter me thinks we need to discuss whether we really need this. Thinking about the current behavior, if PVC isn't found, VolumeBinding PreFilter rejects the pod, the pod goes back to the queue only with `unschedulablePlugins: VolumeBinding`, which means, either way, the queue would determine when to retry this pod only by the volume binding plugin's QHint. In other words, even if the volume binding plugin's QHint returned `Pending` in that case, the queue wouldn't change the behaviour.

So, we have to keep in mind, `Pending` is only useful when _multiple_ plugins are registered at `unschedulablePlugins`, but this pod has to wait for plugin-a's QHint to return `Queue` before retrying the scheduling.

**Comment by sanposhiho** [comment]
@saza-ku 

just ignore my last comment ^. I was mixing up Pending with another status that was proposed but wasn't implemented after all.

**Comment by sanposhiho** [comment]
...but, either way, we need to rethink about if it's appropriate to use `Pending` for PVC-not-found case.

To provide the full context, DRA had a `PodSchedulingContext` that is used to have communication between the resource driver and the scheduler before the structured parameters. And, at that time, DRA had a scenario where the scheduling cycle finished successfully, but needed to hear the opinion from the device driver to determine the final scheduling result.
That's what `Pending`'s documentation describes:
> 	// because in this case, the scheduler decides where the Pod can go successfully,
	// but we need to wait for the external component to do something based on that scheduling result.

So, in other words, DRA's scheduling had to go thru multiple scheduling cycles by design, and we didn't want to put a backoff penalty on those pods because those scheduling failures weren't their fault; rather it's by design that they had to go thru multiple scheduling cycles. We usually put a backoff penalty on pods that have consumed the scheduling cycles because those scheduling cycles are wasted by them. But, on the contrary, the scheduling cycles consumed for DRA pods weren't waste.
That's why `Pending` is allowed to skip a backoff penalty even though they consume the scheduling cycles. 

Then, coming back to the original topic, PVC scenario. The point is whether the consumption of the scheduling cycles is necessary for VolumeBinding or not. And, I think it is not. So, I, again, think we shouldn't change VolumeBinding plugin to return `Pending` by a different reason.

**Comment by sanposhiho** [comment]
In that sense, the past me said _There should be more usecase everywhere_, but I doubt it now. It's for specific, sophisticated, and complex scheduling use cases, and most likely 99% of scheduling plugins in the world don't have to use it.

**Comment by saza-ku** [comment]
Thanks for complementing the context.

I found a case where rejection of the first scheduling cycle is inevitable but it's not necessary.
When a Pod with a generic ephemeral inline volume is created, the pvc is not created so it usually returns `UnschedulableAndUnresolvable` and causes unrequired backoff.

> // The error usually has already enough context ("persistentvolumeclaim "myclaim" not found"),
 // but we can do better for generic ephemeral inline volumes where that situation
 // is normal directly after creating a pod.

https://github.com/kubernetes/kubernetes/blob/1c32094c03f721cd3dce49c9743f1e7f3e564b13/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L324

I thought `Pending` would avoid it but it wouldn't be appropriate based on your comments. Should we leave it or fix it in another way?

**Comment by sanposhiho** [comment]
> When a Pod with a generic ephemeral inline volume is created, the pvc is not created so it usually returns UnschedulableAndUnresolvable and causes unrequired backoff.
...
I thought Pending would avoid it but it wouldn't be appropriate based on your comments. Should we leave it or fix it in another way?

Yeah, that's surely not an appropriate scenario using `Pending`. 
If we want to enhance the scheduling performance on those kind of scenarios, most likely it's better to use `PreEnqueue` so that the pods don't enter the scheduling cycles in the first place if PVCs aren't ready yet.

**Comment by sanposhiho** [comment]
/close

I think we can say there's no in-tree plugin that can utilize `Pending` for now. Closing the issue.

Regarding PVC stuff, I opened https://github.com/kubernetes/kubernetes/issues/129698 so that we can continue the discussion.
